iwinfo: various fixes and improvements

dhewg opened the pull request:
This fixes some bugs I noticed while playing with meshes, usb wifi devices and 6G. See the description of the patches.

This isn't meant to be commited as-is, but rather to the iwinfo repo itself if accepted. This PR is just to make use of the CI setup and allow users to test this more easily.

dhewg committed:
iwinfo: various fixes and improvements

This isn't meant to be commited as-is, but rather to the iwinfo repo
itself if accepted. This PR is just to make use of the CI setup and
allow users to test this more easily.

mj22226 commented:
@dhewg nice work with the frequency/channel conversion on 6g band, channel 25 was 215, now it can be set properly, tested with a mt7921k WiFi card
![Screenshot318](https://user-images.githubusercontent.com/67804477/202780829-480ecf47-e993-4695-977e-ee212d66c0d9.jpg)

dhewg commented:
Looks good, thanks!
What does the 'Hardware:' line look like with `iwinfo` on that card?

dhewg commented:
Alright, so it's already part of devices.txt.
And I'm waiting for that one to arrive ;)

Ansuel commented:
@dhewg i think you already sent this to the mailing list split but just in case, better asking if this was correctly split.

dhewg commented:
I did, one single patch and two set. I updated them and added some more. So this PR replaces them all ;)
And I mentioned this on the ML in a reply to that single patch.

I think this makes it easier for everyone, but let me know if you prefer it the other way around.

dhewg commented:
I converted this to a draft. There're more fixes and cleanup coming...

dhewg committed:
rpcd: iwinfo cleanups and improvements

This isn't meant to be commited as-is, but rather to the rpcd repo
itself if accepted. This PR is just to make use of the CI setup and
allow users to test this more easily.

dhewg commented:
The patches just keep piling up... but they're all small. I'll stop here so it's still reviewable.

Open questions:
- In "nl80211: add support for HE htmodes" I didn't find a way with nl80200 to get if a wiphy is in ax/he mode, so the code currently checks hostapd or wpa_supplicant for that. But there has to be a way to do this in a cleaner way, right?
- Adding USB devices as vendor=product=0 is a little hacky, but so is the current ftd integration. Is that okay?

hauke commented on package/network/utils/iwinfo/patches/0003-fix-Wduplicate-decl-specifier-warnings.patch:
```diff
@@ -0,0 +1,88 @@
++   1 = 80 MHz
++   2 = 160 MHz
++   3 = 80+80 MHz */
++extern const uint16_t vht_chan_width[4];
```
Do you get a compile warning when the size defined here does not match the one defined in the `iwinfo_lib.c` file?

dhewg replied:
I expected that, but for whatever reason I didn't. Which is why I noticed that `-Wall` wasn't used, but even with that I don't get a warning.

hauke commented on package/network/utils/iwinfo/patches/0003-fix-Wduplicate-decl-specifier-warnings.patch:
```diff
@@ -0,0 +1,88 @@
+ 	"HE160"
+ };
+ 
++const char *ht_secondary_offset[4] = {
```
Please make the chars also const. 
```suggestion
+const char * const ht_secondary_offset[4] = {
```

dhewg replied:
Yeah, I noticed that too but didn't do it in the end. I guess consumers will get "discards qualifier" warnings then, which may break builds.
But lemme try if it actually does...

dhewg replied:
Done, but rpcd builds with `-Werror` and thus failed with "discards qualifier". I added a rpcd patch for that, but now we have a dependency, so if we keep it like this both have to be commited or we have a broken build. That's rest builds fine though, at least for my config. Let's see if CI finds more spots...

hauke commented on package/network/utils/iwinfo/patches/0015-nl80211-fix-frequency-channel-conversion-for-the-6G-.patch:
```diff
@@ -0,0 +1,101 @@
+ band
+
+Update using current linux sources, which this was apparently
+based on.
```
If this is copied from some other implementation please mention this in the file.
I think we should also preserve the copyright statements.

dhewg replied:
I did mention it in as a comment, but right you are, added the license and copyright statements too

hauke commented on package/network/utils/iwinfo/patches/0018-nl80211-add-support-for-HE-htmodes.patch:
```diff
@@ -0,0 +1,84 @@
++	if (nl80211_hostapd_query(res ? res : ifname, "ieee80211ax", b, sizeof(b)))
++		he = b[0] == '1';
++	else if (nl80211_wpactl_query(res ? res : ifname, "wifi_generation", b, sizeof(b)))
++		he = b[0] == '6';
```
Where is `b` used?

dhewg replied:
Just for that `bool he`, `b` is just an intermediate since we're getting `char*` results form the socket/config.

hauke replied:
ok, now I get the code.

hauke commented on package/network/utils/iwinfo/patches/0020-lib-fixup-the-string-array-definitions.patch:
```diff
@@ -0,0 +1,55 @@
+  * name constants
+  */
+-const char *IWINFO_CIPHER_NAMES[] = {
++const char *IWINFO_CIPHER_NAMES[IWINFO_CIPHER_COUNT] = {
```
Please change this to const chars:
```suggestion
+const char * const IWINFO_CIPHER_NAMES[IWINFO_CIPHER_COUNT] = {
```

dhewg replied:
done

hauke commented on package/network/utils/iwinfo/patches/0028-utils-add-and-use-iwinfo_format_hwmodes.patch:
```diff
@@ -0,0 +1,96 @@
+"802.11ac/ax/b/g/n" instead of "802.11bgnacax".
+
+Luci currently uses a natural sort order, but that probably doesn't work
+as intended once "eb" is added, so let's do this here.
```
`be` is wifi 7, not `eb`.

dhewg replied:
Heh, right, will fix

dhewg commented:
Thanks for digging through the pile, addressed all your comments!

McGiverGim commented:
@dhewg it seems the signature in the commit is not right:
> Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: Andre Heider <a.heider@gmail.com>')

jow- commented:
I picked most patches for rpcd and iwinfo, omitting the ones changing the ABI. Will follow up with the rest once we bumped iwinfo in 22.03.2 while it is still ABI compatible.

Ansuel commented:
@jow- are we planning to change the ABI? I would love to include the changes for the channel limitation (channel restricted to indoor only and related)

jow- commented:
@Ansuel - ideally not - not yet anyway.

dhewg commented:
Rebased and added some patches to improve this situation: https://github.com/openwrt/luci/issues/6167#issuecomment-1364500296

Ansuel commented:
@dhewg btw jow said that we can finally change the iwinfo ABI so we can merge the ABI breakage changes

Ansuel commented:
@jow- since we need to change ABI good time for the channel limitation (channel restricted to indoor only and related) change?

jow- commented:
@Ansuel - yes, please go ahead with this as well

Ansuel commented:
@dhewg i merged stuff here https://git.openwrt.org/?p=project/iwinfo.git;a=shortlog;h=refs/heads/ABI-changes-20230110 it would be good if you can test this... I'm creating a similar branch for rpcd

dhewg commented:
I'm rebasing my branch for almost two month now, I don't worry about my patches too much, but looks good... except your new freq flags conflict:

```
+#define IWINFO_FREQ_NO_IR		(1 << 0)
+#define IWINFO_FREQ_NO_OUTDOOR		(1 << 1)
```
with
```
#define IWINFO_FREQ_NO_10MHZ		(1 << 0)
#define IWINFO_FREQ_NO_20MHZ		(1 << 1)
```

I guess you want the next free bits, so <<7 and <<8?

Ansuel commented:
@dhewg they don't as they are put in restricted_flags

dhewg commented:
Right, saw that just after posting ;)
But can't we add those bits to the existing `flags` instead? All other *_FREQ_NO_* bits are in there and there're enough free bits

Ansuel commented:
I mean from nl80211 https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/nl80211.h#L4132

Ansuel commented:
@dhewg on second tought checking the kernel source i decided to follow your advice and merge everything in the flags space... rpcd iwinfo module should always parse these flag and the order is not really important.... 

Waiting for your test and then I will merge them in the master branch. (the new branch are:
- https://git.openwrt.org/?p=project/iwinfo.git;a=shortlog;h=refs/heads/ABI-changes-20230110-2
- https://git.openwrt.org/?p=project/rpcd.git;a=shortlog;h=refs/heads/ABI-changes-20230110-2

dhewg commented:
Alright, I think it fits pretty well, since all the flags we use are basically negative ones - restrictions.

Btw, is anything using `restricted`? If I read that correctly it's now getting set if INDOOR_ONLY is. That's a restriction, alright, but indoor usage is pretty much the standard - not that anything in tree limits the channels because of that now?

Ansuel commented:
@dhewg from luci side restricted is used as a bool to check if a specific channel have any kind of restriction... Currently it was set to NO_IR but this is not expanded for the indoor only limitation.

Luci will then be updated to add some warning or remove the channel from the channel list based on the additional info from no_outdoor or no_ir and these check are triggered by the restricted bool.

Ansuel commented:
btw I was also checking these fixes 
https://patchwork.ozlabs.org/project/openwrt/patch/20210517080518.26077-1-wojciech.jowsa@gmail.com/
https://patchwork.ozlabs.org/project/openwrt/patch/20210507113856.28810-1-wojciech.jowsa@gmail.com/

dhewg commented:
So if a channel now gets restricted due to INDOOR_ONLY luci still accepts it? Just making sure we're not introducing regression without further luci changes. IIRC 6g is limited to indoor in EU. Please don't wipe all my channels :P

dhewg commented:
And yeah, I pointed @jow- to those 2 patchwork entries. He had something else in mind to solve the issue IIRC.

Ansuel commented:
Yes with current luci you channel will be wiped but a pr for luci is already there that should fix the situation https://github.com/openwrt/luci/pull/5695

I just need to update it with recent changes

dhewg commented:
Indeed, all 6g channels are gone, hehe.

That luci PR conflicts with my patches, so I used this for now instead:
```
diff --git a/iwinfo_nl80211.c b/iwinfo_nl80211.c
index 8bec8ae..d960ffa 100644
--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -3037,9 +3037,8 @@ static int nl80211_get_freqlist_cb(struct nl_msg *msg, void *arg)
 					e->mhz = nla_get_u32(freqs[NL80211_FREQUENCY_ATTR_FREQ]);
 					e->channel = nl80211_freq2channel(e->mhz);
 
-					e->restricted = ((freqs[NL80211_FREQUENCY_ATTR_NO_IR] &&
-							 !freqs[NL80211_FREQUENCY_ATTR_RADAR]) ||
-							 freqs[NL80211_FREQUENCY_ATTR_INDOOR_ONLY]) ? 1 : 0;
+					e->restricted = freqs[NL80211_FREQUENCY_ATTR_NO_IR] &&
+							 !freqs[NL80211_FREQUENCY_ATTR_RADAR];
 
 					if (freqs[NL80211_FREQUENCY_ATTR_NO_IR] &&
 					    !freqs[NL80211_FREQUENCY_ATTR_RADAR])
```

So it all looks good to me! If the luci PR is ready this should be ready to go in. In not maybe merge it without using INDOOR_ONLY for restricted, at least for now?

Ansuel commented:
The luci pr needs some additional talk with jow to understand about that warning box

Ansuel commented:
> So it all looks good to me! If the luci PR is ready this should be ready to go in. In not maybe merge it without using INDOOR_ONLY for restricted, at least for now?

it's an option but lets see if today we manage to finish the discussion for the luci pr... If that's the case we can just merge everything as is and bump iwinfo/rpcd and merge luci changes at the same time

dhewg commented:
Sounds good to me.

Maybe it even makes sense to add an in/outdoor mode to configure? IIRC there may be additional power restrictions if outdoor.

Ansuel commented:
> Maybe it even makes sense to add an in/outdoor mode to configure? IIRC there may be additional power restrictions if outdoor.

can you point me where nl80211 expose lower power emission if used outdoor?

dhewg commented:
I haven't used it, only read about it. And maybe I'm confusing things? Let's just ignore that :)

dhewg commented:
Yeah, there's quite some more info that iwinfo could expose. But is there a use case for per scanentry country codes? (It also only works with nl80211, not the supplicant code path - in the very same way we encountered with HT/VHT channel widths).

W.r.t. lua, I didn't update that API for the new stuff I added on the C side, lua will slowly fade out of the picture anyway.

But if you're into interesting patches, here's 4 more for you :)
https://github.com/dhewg/iwinfo/commits/master

dhewg commented:
Looking at your luci PR, we could just drop `restricted`. It's just a bool based on chosen `flags` bits anyway.
Users interested in some chosen bits could just test for those instead of the ad-hoc bool. That'll make the luci code less awkward too.

dhewg commented:
Sorry, yet another request for the flags thingy, just looked at your rpcd changes more closely now:

This isn't even compile tested, but could you add something like this?
```
diff --git a/include/iwinfo.h b/include/iwinfo.h
index 71e3dd2..441d090 100644
--- a/include/iwinfo.h
+++ b/include/iwinfo.h
@@ -71,12 +71,14 @@
 #define IWINFO_FREQ_NO_HE      (1 << 6)
 #define IWINFO_FREQ_NO_IR      (1 << 7)
 #define IWINFO_FREQ_NO_OUTDOOR     (1 << 8)
+#define IWINFO_FREQ_FLAGS_COUNT        9
 
 extern const char * const IWINFO_80211_NAMES[IWINFO_80211_COUNT];
 extern const char * const IWINFO_BAND_NAMES[IWINFO_BAND_COUNT];
 extern const char * const IWINFO_CIPHER_NAMES[IWINFO_CIPHER_COUNT];
 extern const char * const IWINFO_KMGMT_NAMES[IWINFO_KMGMT_COUNT];
 extern const char * const IWINFO_AUTH_NAMES[IWINFO_AUTH_COUNT];
+extern const char * const IWINFO_FREQ_FLAGS_NAMES[IWINFO_FREQ_FLAGS_COUNT];
 
 
 enum iwinfo_opmode {
diff --git a/iwinfo_lib.c b/iwinfo_lib.c
index 9f787ac..313e58c 100644
--- a/iwinfo_lib.c
+++ b/iwinfo_lib.c
@@ -95,6 +95,18 @@ const char * const IWINFO_HTMODE_NAMES[IWINFO_HTMODE_COUNT] = {
    "HE160"
 };
 
+const char * const IWINFO_FREQ_FLAGS_NAMES[IWINFO_FREQ_FLAGS_COUNT] = {
+   "no_10mhz",
+   "no_20mhz",
+   "no_ht40+",
+   "no_ht40-",
+   "no_80mhz",
+   "no_160mhz",
+   "no_he",
+   "no_ir",
+   "no_outdoor",
+};
+
 const char * const ht_secondary_offset[4] = {
    "no secondary",
    "above",
```

That's in line what all the other flags do, and it allows you do to this in rpcd:
`rpc_iwinfo_add_bit_array("flags", f->flags, IWINFO_FREQ_FLAGS_NAMES, IWINFO_FREQ_FLAGS_COUNT, true, 0);`
Instead of adding bits manually one by one.

That should give you not just all the bits in a nice ubus array, but it also means that one just adds future flags with its name to iwinfo and they'll pop up in the ubus output without further rpcs code changes automagically.

And that in return means you could just use that as ready-to-use json arrays in luci instead of the available var you added there. Which in the end makes the removal of `restricted` even more obvious :)

Ansuel commented:
@dhewg it's not a bad idea... Actually I like it... Just I think we should make it a big enum instead of adding a define for the count... Less error prone.

dhewg commented:
I totally agree with the enum argument, but the first 6 bits have been there for years, I can only guess if changing the defines to an enum will break compilation of something? So I'd just err on the side of caution and just add stuff instead of changing it...

...for now. Once the iwinfo patch frequency settles we can go wild for iwinfo v2 or iwinfo.uc or whatever shape it's going to be.

And I guess you mean removing your handle from the SoB? Will do, but for the record, I didn't add that, I used curl|git am. And EDONTCARE about Co-Developed :)

dhewg commented:
Staring a bit at this again, I believe we shouldn't add the indoor flag to `restricted`, since this will just regress users without further changes. Instead, let users pick the bits from `flags` as they which. With that we can merge this independent of other PRs. And once all in-tree users pick their bits, we can remove `restricted` at a later point.

Which is exactly what I did, updated my branch and this PR with the very same patches (which makes the patch links above outdated).

@Ansuel  I think that's a sane path forward, what do you think?

dhewg commented:
Fixed a bug in the supplicant patch (different phy/net order after a reboot triggered this).
You may have noticed, there're some mediatek id patches in here now, I can open a new PR for that if that's desired.

Ansuel commented:
Sorry just got time to recheck this...

> You may have noticed, there're some mediatek id patches in here now, I can open a new PR for that if that's desired.

Considering they are things not related to this and that can be merged right away I would ask you to send them to mailing list so we can merge them right away instead of keeping the commit number too big.

dhewg commented:
I went ahead and changed it to IWINFO_FREQ_INDOOR_ONLY.
And I send the MT additions to the mailing list, so there're no longer in here.

I hope that's okay now and we can finally merge this

Ansuel commented:
@dhewg sorry for the delay... yes it's not a bad idea

dhewg commented:
np, iwinfo is not exactly fun, but it's definitely more fun than llvm :P

dhewg commented:
ah damn, forgot to change the string to INDOOR_ONLY, fixed

Ansuel commented:
Btw I reworked the header to enum logic and end up with a smaller package... The magic of gcc optimization...

Using ./script/size-compare.sh
Your patches (with the additional devices stuff)
```
+738    2943    2205    libiwinfo-data
+319    7490    7171    iwinfo
```

My patches (still have to publish)
```
+738    2943    2205    libiwinfo-data
+251    7422    7171    iwinfo
```

Since we have that problem with branch in iwinfo I will publish the changes here.

Ansuel commented:
@dhewg bump iwinfo and rpcd in master. Closing this.

dhewg commented:
Woot,thanks!

dhewg commented:
Dunno how to really verify it, but do packages depending on libiwinfo need a manual bump? Or is the already bumped IWINFO_ABI_VERSION sufficient for that?

Ansuel commented:
They should all allign to the abiversion I assume... We should check a new
image if things works correctly

Il Sab 21 Gen 2023, 16:51 dhewg ***@***.***> ha scritto:

> Dunno how to really verify it, but do packages depending on libiwinfo need
> a manual bump? Or is the already bumped IWINFO_ABI_VERSION sufficient for
> that?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/11280#issuecomment-1399276842>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQWBRDF43D4PULYOWITWTQAXDANCNFSM6AAAAAASEOCHL4>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

dhewg commented:
There won't be an issue with older packages using a newer libiwinfo, I think:
`$ fgrep libiwinfo /usr/lib/opkg/*`

But eventually we want all apps to be rebuilt against the new one, I'm not sure if additional steps are required. I build everything myself

cedbouchet commented:
Hello, 
Sorry to ask but I have exactly the same issue on "OpenWrt 22.03.5 r20134-5f15225c1e" on a Xiaomi AX3022 router - Encryption None for Mesh Point. Is my version include this patch of libiwinfo? If it is not the case; could you please help me to understand  how to apply this patch. I review the following page but I am not really sure on how to download and apply this patch: 
https://openwrt.org/docs/guide-developer/toolchain/use-patches-with-buildsystem

Thanks for support, 
Cédric

tapper82 commented:
@cedbouchet https://forum.openwrt.org/
